### DISCUSSION TITLE
StrimziPodSet don't re-scale kafka pod after draining node

### DISCUSSION THREAD

**Discussion by Rabzou** [strimzi/strimzi-kafka-operator#8447] [Q&A] [answered]
Hi Strimzi team,
I'm using Strimzi 0.34.0 with kafka 3.4.0.
I try to drain node and test the scheduling of kafka pods.
So with StatefulSet, (by disabling StrimziPodSet feature gate) i have no problem. The kafka pod is correctly delete, and scaled up after the drain.

But with StrimziPodSet, the kafka pod is well deleted, but it don't re-scale.
I get this error in the operator logs:

```
2023-04-26 15:48:14 ERROR StrimziPodSetController:374 - Reconciliation #906(watch) StrimziPodSet(brice-project/myoffice-myapp-strimzi-kafka): StrimziPodSet myoffice-myapp-strimzi-kafka in namespace brice-project reconciliation failed
12718io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://172.30.0.1:443/api/v1/namespaces/brice-project/pods. Message: Internal error occurred: failed calling webhook "sidecar-injector.istio.io": failed to call webhook: Post "https://istiod-basic-install.brice-project-istio.svc:443/inject?timeout=10s": no endpoints available for service "istiod-basic-install". Received status: Status(apiVersion=v1, code=500, details=StatusDetails(causes=[StatusCause(field=null, message=failed calling webhook "sidecar-injector.istio.io": failed to call webhook: Post "https://istiod-basic-install.brice-project-istio.svc:443/inject?timeout=10s": no endpoints available for service "istiod-basic-install", reason=null, additionalProperties={})], group=null, kind=null, name=null, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Internal error occurred: failed calling webhook "sidecar-injector.istio.io": failed to call webhook: Post "https://istiod-basic-install.gre05-br...
12719at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:238) ~[io.fabric8.kubernetes-client-api-6.3.1.jar:?]
12720at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:536) ~[io.fabric8.kubernetes-client-6.3.1.jar:?]
12721at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:570) ~[io.fabric8.kubernetes-client-6.3.1.jar:?]
12722at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:554) ~[io.fabric8.kubernetes-client-6.3.1.jar:?]
```

Am I missing a StrimziPodSet parameter to ensure Scheduling and pod management?

**Comment by scholzj** [comment]
The error suggests the problem is with your Istio webhook and due to it Kubernetes cannot create the Pod as requested by Strimzi:

```
Internal error occurred: failed calling webhook "sidecar-injector.istio.io": failed to call webhook: Post "https://istiod-basic-install.brice-project-istio.svc:443/inject?timeout=10s": no endpoints available for service "istiod-basic-install"
```

So I guess you should remove the Istio admission controller or fix it to work. It is not clear what your intention is - Strimzi does not support Istio and its sidecars - although I think some people made it to work.

It is also not obvious why would the StatefulSets pods not be blocked by the same error, but it is hard to say without knowing your environment.

**Reply by Rabzou** [reply]
Thanks for the quick reply!
I wanted to test a drain node. So if I do it, the pod must be rescale in another node.

My environment before the drain:

```
ocp-rosiscky-qtw46-worker-wxzqs
  brice-project-istio                    istiod-basic-install-7f77d7cf4f-j96np                      10m (0%)      0 (0%)      128Mi (0%)       0 (0%)         119m
  brice-project-istio                    prometheus-687f786f8b-shdcv                                20m (0%)      0 (0%)      256Mi (0%)       0 (0%)         119m
  brice-project                          brice-project-istio-ingressgateway-5456fc9c4d-5sp4r       1 (6%)        2 (12%)     1Gi (1%)         1Gi (1%)       119m
  brice-project                          myoffice-mariadb-myapp-supervisor-0                                    250m (1%)     1 (6%)      1Gi (1%)         2Gi (3%)       119m
  brice-project                          myoffice-operator-sandrine-579db8c869-znfsd                  250m (1%)     500m (3%)   100Mi (0%)       256Mi (0%)     119m
  brice-project                          myoffice-server-86d5bcdc59-g8nqn                            300m (1%)     2 (12%)     1Gi (1%)         1536Mi (2%)    119m
  brice-project                          myoffice-mariadb-operator-brice-project-7fdb59df9f-czprc    100m (0%)     100m (0%)   100Mi (0%)       200Mi (0%)     119m
  brice-project                          myoffice-myapp-59d9df95bc-lvjfz                  200m (1%)     1 (6%)      1Gi (1%)         2Gi (3%)       119m
  brice-project                          myoffice-myapp-strimzi-kafka                                   800m (5%)     1 (6%)      2Gi (3%)         2Gi (3%)       118m
```

All pods are correctly rescaled elsewhere, except kafka pod (if using StrimziPodSet)
If you plan to remove StatefulSet management in version 0.35.0, I must find a way to configure my environment with StrimziPodSet and the drain. But I can't remove istio.

You say some people make it to work with strimzi. 
Did you know how?

**Reply by scholzj** [reply]
Why can't you remove Istio? It does not support Kafka, so it gives you probably only very little value. But in any case, you need to look at things in the right order -> the immediate problem you have is the problem with the Istio webhook which is what is causing the error. So you need to find out why that does not work and does not have any endpoints.

**Reply by Rabzou** [reply]
The strimzi's pods didn't have istio sidecar. But are in the same namespace (brice-project) as other pods.
This namespace is member of the servicemesh.

Right now, I made a test by moving them in another namespace, and I didn't get the error (even with StrimziPodSet).
So, why in previous setup Strimzi get this istio webhoock ?

**Reply by scholzj** [reply]
> So, why in previous setup Strimzi get this istio webhoock ?

I'm afraid I do not know because I have only a little knowledge about your environment. So that is probably a question on your _Istio installation_. I know only a little about Istio - so I have no idea why it was now available etc.

**Reply by Rabzou** [reply]
I don't know much about istio and service mesh, but I can describe my set up:
I have a redhat service mesh operator installed on all namespaces. And in this service mesh member roll, the namespace "brice-project" is declared.
On this namespace I have a NetworkPolicy that accept all traffic.

I guess through this mechanism, all pods in this namespace must pass by istio...
I'll take a look on the installation.

**Reply by Rabzou** [reply]
I was looking for a way to leave kafka outside istio, and I found that:
https://github.com/strimzi/strimzi-kafka-operator/issues/817

It's already possile to add annotations to bypass istio.

**Reply by scholzj** [reply]
You can add annotations, sure. But I _think_ the annotations will be interpreted by the Istio Admission controller. In the error you had at the beginning, the Istio Admission Controller did not seem to work. So I'm not sure the annotation will help.

**Reply by Rabzou** [reply]
During the drain, some pods are removed from a cluster and rescale on another.
So I guess the istiod pod is also moving... maybe during his move, StrimziPodSet can't reach him. that's why we get the error of the begining. So it stop, and don't scale kafka anymore.

Maybe in StatefulSet mode, there is much more retry operation, or a greater timeout. So when the istiod come back it can scale kafka pods. That's why we don't have error in this mode.

What do you think about this assumption ?

**Reply by scholzj** [reply, marked as answer]
That would be possible -> this would be a known problem with admission controllers. You should normally either have it HA or have it configured to let Kubernetes [ignore it if it is not available](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#failure-policy). https://github.com/strimzi/strimzi-kafka-operator/pull/8464 should add resync (periodical retry) in the next Strimzi version. But I think fixing the Istio controller should be done regardless of that.

**Reply by Rabzou** [reply]
Tested with strimzi 0.35.1, and there is no problem anymore.
Thanks @scholzj !
